video_core: Replace gl_shader_decompiler with an IR based decompiler

ReinUsesLisp opened the pull request:
This PR aims to solve two issues with one big bullet:
* to have an API agnostic decompiler and
* to be able to do code analysis on shaders (like constant propagation for LDG).

I know this is a pain to review (as it was also to write), so to split the work here's a check list of reviewed instructions (to ease the track):

- [ ] MOV
- [ ] FMUL
- [ ] FADD
- [ ] MUFU
- [ ] FMNMX
- [ ] RRO
- [ ] HADD2 and HMUL2
- [ ] HADD2 and HMUL2 (immediate)
- [ ] MOV32_IMM
- [ ] FMUL32_IMM
- [ ] FADD32_IMM
- [ ] IADD
- [ ] IADD3
- [ ] ISCADD
- [ ] POPC
- [ ] SEL
- [ ] LOP
- [ ] LOP3
- [ ] IMNMX
- [ ] LEA
- [ ] IADD32I
- [ ] LOP32I
- [ ] BFE
- [ ] BFI
- [ ] I2I
- [ ] I2F
- [ ] F2F
- [ ] F2I
- [ ] FFMA
- [ ] FSET
- [ ] FSETP
- [ ] HSET
- [ ] HSETP
- [ ] HFMA2
- [ ] ISET
- [ ] ISETP
- [ ] LD_A
- [ ] LD_C
- [ ] LD_L
- [ ] ST_A
- [ ] ST_L
- [ ] TEX
- [ ] TEXS
- [ ] TLD4
- [ ] TLD4S
- [ ] TXQ
- [ ] TMML
- [ ] TLDS
- [ ] EXIT
- [ ] KIL
- [ ] MOV_SYS
- [ ] BRA
- [ ] SSY and SYNC
- [ ] PBK and BRK
- [ ] IPA
- [ ] OUT
- [ ] ISBERD
- [ ] DEPBAR
- [ ] PSETP
- [ ] PSET
- [ ] R2P
- [ ] SHR
- [ ] SHL
- [ ] VMAD
- [ ] VSETP
- [ ] XMAD

For this to be in a mergeable state I have to implement the remaining and to have a replacement for #1603. There's also some stuff that's wrong at the moment (like array indexes in textures).

Thanks to Ryujinx's IR that served as a base to implement this (there are major differences but without it I couldn't have written it).

lioncash commented on src/video_core/shader/shader_ir.h:
```diff
@@ -0,0 +1,764 @@
+    /// Creates a f32 immediate
+    Node Immediate(f32 value) {
+        // TODO(Rodrigo): Replace this with bit_cast when C++20 releases
+        return Immediate(*reinterpret_cast<const u32*>(&value));
```
You can do:

```cpp
u32 integral;
std::memcpy(&integral, &value, sizeof(u32));
```

to do the same thing in the meantime without invoking undefined behavior.

lioncash commented:
Note that `assert.h` appears to be included in many headers that don't actually make use of it.

lioncash commented on src/video_core/shader/decode.cpp:
```diff
@@ -0,0 +1,199 @@
+using Tegra::Shader::OpCode;
+
+/// Merges exit method of two parallel branches.
+constexpr ExitMethod ParallelExit(ExitMethod a, ExitMethod b) {
```
This and `IsSchedInstruction` should be placed within an anonymous namespace, otherwise they don't have internal linkage

lioncash commented on src/video_core/shader/shader_ir.h:
```diff
@@ -0,0 +1,764 @@
+                              Node op_c, Node imm_lut);
+
+    template <typename... T>
+    inline Node Operation(OperationCode code, const T*... operands) {
```
You don't need to mark member functions `inline` if they're defined in the class declaration, they're inline by default.

lioncash commented on src/video_core/shader/shader_ir.h:
```diff
@@ -0,0 +1,764 @@
+/// Commentary, can be dropped
+class CommentNode final {
+public:
+    explicit CommentNode(const std::string& text) : text{text} {}
```
```suggestion
    explicit CommentNode(std::string text) : text{std::move(text)} {}
```

lioncash commented on src/video_core/shader/shader_ir.h:
```diff
@@ -0,0 +1,764 @@
+    }
+
+private:
+    const std::string text;
```
This should probably be non-const, otherwise it makes the class non-movable, an will always construct a copy, even if `std::move` is used on an instance of it, ditto for all the other nodes above this comment.

lioncash commented on src/video_core/shader/shader_ir.h:
```diff
@@ -0,0 +1,764 @@
+/// Global memory node
+class GmemNode final {
+public:
+    explicit GmemNode(Node address) : address{address} {}
```
Was this supposed to be `constexpr` as well? (other nodes above have it)

lioncash commented on src/video_core/shader/shader_ir.h:
```diff
@@ -0,0 +1,764 @@
+    }
+
+    u32 GetIndex() const {
+        return static_cast<u32>(index);
```
any reason why `std::size_t` isn't just the return type?

ReinUsesLisp replied:
It was mostly because it's usage was as an `u32` from where the value was being queried. Changed it to `std::size_t` for consistency.

lioncash commented on src/video_core/shader/shader_ir.h:
```diff
@@ -0,0 +1,764 @@
+
+class ShaderIR final {
+public:
+    explicit ShaderIR(const ProgramCode& program_code, u32 main_offset)
```
```suggestion
    explicit ShaderIR(ProgramCode program_code, u32 main_offset)
```

lioncash commented on src/video_core/shader/shader_ir.h:
```diff
@@ -0,0 +1,764 @@
+class ShaderIR final {
+public:
+    explicit ShaderIR(const ProgramCode& program_code, u32 main_offset)
+        : program_code{program_code}, main_offset{main_offset} {
```
```suggestion
        : program_code{std::move(program_code)}, main_offset{main_offset} {
```

ReinUsesLisp commented on src/video_core/shader/shader_ir.cpp:
```diff
@@ -0,0 +1,408 @@
+
+/*static*/ OperationCode ShaderIR::SignedToUnsignedCode(OperationCode operation_code,
+                                                        bool is_signed) {
+    if (is_signed) {
```
It's there to avoid having an `if (is_signed)` on the caller.

heapo commented on src/video_core/shader/decode/memory.cpp:
```diff
@@ -102,6 +103,44 @@ u32 ShaderIR::DecodeMemory(BasicBlock& bb, u32 pc) {
+        UNIMPLEMENTED_IF_MSG(instr.tex.UsesMiscMode(TextureMiscMode::AOFFI),
+                             "AOFFI is not implemented");
+
+        if (instr.tex.UsesMiscMode(TextureMiscMode::NODEP)) {
```
First, awesome approach, obviously lots of potential here.

Honest question, do you think it makes sense to bubble NODEP down into the IR (and then ignore it in the GLSL backend)? You downgraded the UNIMPLEMENTED_IF to a LOG_WARNING so curious what you thought.

(Either way probably fine for a separate PR to tackle all the NODEP stuff).

heapo commented on src/video_core/shader/decode/memory.cpp:
```diff
@@ -0,0 +1,618 @@
+            }
+            std::generate(dest.begin() + dest_elem, dest.end(), [&]() { return GetRegister(RZ); });
+
+            bb.push_back(Operation(OperationCode::AssignComposite, std::move(meta), texture,
```
Am I interpreting correctly that this zeroes out the disabled components? That's a change in behavior so wanted to double check if intentional (maybe a good fix..).

heapo commented on src/video_core/shader/decode/memory.cpp:
```diff
@@ -0,0 +1,618 @@
+
+    std::vector<Node> params = std::move(coords);
+
+    if (process_mode != TextureProcessMode::None) {
```
Quick note to update to the changes made in https://github.com/yuzu-emu/yuzu/pull/1909 re: LOD selection

heapo commented on src/video_core/shader/decode/memory.cpp:
```diff
@@ -0,0 +1,618 @@
+        const u64 depth_register = instr.gpr20.Value() + (lod_bias_enabled ? 1 : 0);
+        coords.push_back(GetRegister(depth_register));
+    }
+    if (is_array) {
```
(Also a change from https://github.com/yuzu-emu/yuzu/pull/1909/files), array index belongs before depth comparison value (also GetTexsCode).

heapo commented on src/video_core/shader/decode/memory.cpp:
```diff
@@ -0,0 +1,618 @@
+            Operation(OperationCode::ICastFloat, NO_PRECISE, GetRegister(array_register)));
+    }
+    // Fill ignored coordinates
+    while (coords.size() < total_coord_count) {
```
Nice refactor with ValidateAndGetCoordinateElement. For the special 1D.DC case, it looks like this will add the '0' noop coordinate to the end of params? It should go as the y value, before the depth compare value.

heapo commented on src/video_core/shader/decode/memory.cpp:
```diff
@@ -0,0 +1,618 @@
+        break;
+    }
+    case OpCode::Id::TLD4: {
+        ASSERT(instr.tld4.texture_type == Tegra::Shader::TextureType::Texture2D);
```
Just left-over debugging code?

ReinUsesLisp replied:
It was like that on the original code.

heapo replied:
Ah I see, those lines were removed in https://github.com/yuzu-emu/yuzu/pull/1862, it looks the rest of this function (probably) needs to be rebased on that

heapo commented on src/video_core/shader/decode/memory.cpp:
```diff
@@ -0,0 +1,618 @@
+            num_coordinates = 2;
+            texture_type = Tegra::Shader::TextureType::Texture2D;
+        }
+        params.push_back(Immediate(static_cast<u32>(instr.tld4.component)));
```
In existing code this is only done for non-DC case, just checking intentional?

heapo commented on src/video_core/shader/decode/memory.cpp:
```diff
@@ -0,0 +1,618 @@
+
+                ++dest_elem;
+            }
+            std::generate(dest.begin() + dest_elem, dest.end(), [&]() { return GetRegister(RZ); });
```
As as earlier comment just checking you believe it's correct to zero out the disabled registers

ReinUsesLisp replied:
Just it pointing as unresolved. I have to take a look.

heapo commented on src/video_core/shader/glsl_decompiler.cpp:
```diff
@@ -0,0 +1,1398 @@
+    }
+
+    std::string F4TextureGather(Operation operation) {
+        return GenerateTexture(operation, "textureGather", "int");
```
For DC gathers, the 'extra' parameter type should be float.

ReinUsesLisp commented:
All instructions that are implemented in master do so here. I'll be addressing feedback and trying to fix issues that I find (like OS X's build failing).

Hexagon12 commented:
Please rebase the PR.
